### PR TITLE
Restrict EdDSA private-key supported types

### LINKED ISSUE #188: Restrict types supported by functions in the EdDSA Poseidon package
**Describe the improvement you're thinking about**

Currently, functions such as `deriveSecretScalar`, `derivePublicKey` and `signMessage` accept different types of private keys. BigNumberish can be: `bigint`, `number`, `string`, hexadecimal (string) or Buffer.  Although supporting and converting many types reduces the dev's effort in certain contexts, it may also lead to ambiguity if a type is not correctly identified. For example, a hexadecimal string without '0x' is identified as plain text or as a bigint (if it contains only digits) depending on its value, and inconsistency remains between the length of an accepted arbitrary text (no limit at the moment) and the value of a bigint or a number, which have different size limits.

**Describe alternatives you've considered**

In general, the right choice for determining how many and which types to support depends on the case. Prioritizing simplicity and accepting only one type (e.g. Buffer) may be the right choice if you want to make your code more efficient and eliminate ambiguity in the interpretation of types. Making the code more flexible and adaptable may be a better choice in the case of prioritizing the user experience.

This is an implementation of an algorithm where performance matters, but at the same time the aim is to give devs an optimal experience. An intermediate solution might be to reduce the number of types to 1 or 2, and provide utilities to convert any values before passing them to library functions.

**Additional context**

Originated from discussion in https://github.com/privacy-scaling-explorations/zk-kit/pull/178#discussion_r1509782449.

### PR DESCRIPTION
**Pull request by cedoor**
## Description

Breaking change ⚠️ 

Types supported for the EdDSA private key have been reduced to 3: Buffer, Uint8Array, and string, i.e. buffers or text.

## Related Issue(s)

Closes #188

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] My changes generate no new warnings
-   [x] I have run `yarn prettier` and `yarn lint` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/eddsa-poseidon/src/eddsa-poseidon.ts [hunk 1]
```diff
@@ -26,12 +31,15 @@ import { hash as blake, checkMessage, checkPrivateKey, isPoint, isSignature, pru
  * and converting it into a little-endian integer. The use of the secret scalar streamlines the public key generation
  * process by omitting steps 1, 2, and 3 as outlined in RFC 8032 section 5.1.5, enhancing circuit efficiency and simplicity.
  * This method is crucial for fixed-base scalar multiplication operations within the correspondent cryptographic circuit.
+ * The private key must be an instance of Buffer, Uint8Array or a string. If you want to pass a bigint, a number or a
```
Comment (artwyman) [inline packages/eddsa-poseidon/src/eddsa-poseidon.ts:34]: Be explicit about how a string is interpreted in this case.  I.e. the string is used as a set of raw bytes (in UTF-8, I believe) as input to key generation.  It's not interpreted as hex, or decimal, or anything else.

Also probably be explicit in saying the input can be any size, and all bytes will be used to generate entropy (vs. other inputs where a Buffer is expected to represent a number of a specific size).

Comment (artwyman) [reply packages/eddsa-poseidon/src/eddsa-poseidon.ts:34]: Same comment in all the other functions which take private key as input, so I won't duplicate this comment.

**Review by artwyman — APPROVED** [review]
Looks good functionally, but see my comment for a suggested clarification to docs.

**Review by 0xjei — APPROVED** [review]

File: packages/eddsa-poseidon/src/eddsa-poseidon.ts (hunk 1 above)
Comment (cedoor) [reply packages/eddsa-poseidon/src/eddsa-poseidon.ts]: Thanks! I updated them 👍🏽

**Pull request merged**